cloud/google/compute/reconciler.go

```diff
@@ -0,0 +1,230 @@
+	return &Reconciler{
+		Known: expected,
+	}
+}
+
+var model map[int]cloud.Resource
+
+// Init is used to create the sdk and add this to the resources.
+func (r *Reconciler) Init() error {
+	sdk, err := googleSDK.NewSdk()
+	if err != nil {
+		return err
+	}
+
+	resources.Sdk = sdk
+	model = ClusterModel(r.Known)
+	return nil
+}
+
+// GetActual is used to create a representation of the cluster existing on the cloud provider.
+func (r *Reconciler) GetActual() (*cluster.Cluster, error) {
+	actualCluster := newClusterDefaults(r.Known)
+	for i := 0; i < len(model); i++ {
+		resource := model[i]
+		actualResource, err := resource.Actual(r.Known)
+		if err != nil {
+			return nil, err
+		}
+		actualCluster, err = resource.Render(actualResource, actualCluster)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return actualCluster, nil
+}
+
+// GetExpected is used to create a representation of the cluster expected on the cloud provider.
+func (r *Reconciler) GetExpected() (*cluster.Cluster, error) {
+	expectedCluster := newClusterDefaults(r.Known)
+	for i := 0; i < len(model); i++ {
+		resource := model[i]
+		expectedResource, err := resource.Expected(r.Known)
+		if err != nil {
+			return nil, err
+		}
+		expectedCluster, err = resource.Render(expectedResource, expectedCluster)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return expectedCluster, nil
+}
+
+func cleanUp(cluster *cluster.Cluster, i int) error {
+	logger.Warning("--------------------------------------")
+	logger.Warning("Attempting to delete created resources!")
+	logger.Warning("--------------------------------------")
+	for j := i - 1; j >= 0; j-- {
+		var err error
+		resource := model[j]
+		createdResource := createdResources[j]
+		_, err = resource.Delete(createdResource, cluster)
+		if err != nil {
+			j, err = destroyI(err, j)
+			if err != nil {
+				return err
+			}
+			continue
+		}
+	}
+	return nil
+}
+
+var createdResources = make(map[int]cloud.Resource)
+
+// Reconcile is used to call all function to create a cluster on the cloud provider
+func (r *Reconciler) Reconcile(actualCluster, expectedCluster *cluster.Cluster) (*cluster.Cluster, error) {
+	newCluster := newClusterDefaults(r.Known)
+
+	for i := 0; i < len(model); i++ {
+		if sigCaught {
+			cleanUp(newCluster, i)
+			os.Exit(1)
+		}
+
+		c := make(chan os.Signal, 1)
+		signal.Notify(c, os.Interrupt)
+
+		go handleCtrlC(c)
+
+		resource := model[i]
+		expectedResource, err := resource.Expected(expectedCluster)
+		if err != nil {
+			return nil, err
+		}
+		actualResource, err := resource.Actual(actualCluster)
+		if err != nil {
+			return nil, err
+		}
+		appliedResource, err := resource.Apply(actualResource, expectedResource, newCluster)
+		if err != nil {
+			logger.Critical("Error during apply! Attempting cleaning: %v", err)
+			err = cleanUp(newCluster, i)
+			if err != nil {
+				logger.Critical("Failure during cleanup! Abandoned resources!")
+				return nil, err
+			}
+			return nil, nil
```

Comment:
We're already in error handling code here -- shouldn't we still return the error rather than `nil, nil`, even if cleanup was successful?